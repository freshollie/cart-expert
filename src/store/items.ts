import {
  createSelector,
  createSlice,
} from "@reduxjs/toolkit";
import { RootState } from ".";
import { Item } from "../types";

const ITEMS: readonly Item[] = [
  {
    id: "34583",
    name: "Beans",
    price: {
      amount: 50,
      divisor: 100,
    },
  },
  {
    id: "39845",
    name: "Coke",
    price: {
      amount: 70,
      divisor: 100,
    },
  },
  {
    id: "32844",
    name: "Oranges",
    price: {
      amount: 199,
      divisor: 100,
      perKg: true,
    },
  },
  {
    id: "34242",
    name: "Bananas",
    price: {
      amount: 199,
      divisor: 100,
      perKg: true,
    },
  },
];

export default createSlice({
  name: "items",
  initialState: ITEMS,
  reducers: {},
});

export const selectItemsLookup = createSelector(
  (state: RootState) => state.items,
  (items) =>
    items.reduce(
      (map, item) => ({ ...map, [item.id]: item }),
      {} as Record<string, Item>
    )
);

export const selectAllItems = (state: RootState) => state.items;

export const selectItembyId = createSelector(
  selectItemsLookup,
  (_: RootState, { id }: { id: string }) => id,
  (map, id) => map[id]
);
