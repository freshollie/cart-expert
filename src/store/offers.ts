import { createSelector, createSlice } from "@reduxjs/toolkit";
import { create } from "domain";
import { RootState } from ".";
import { Offer } from "../types";

// Offers are in a list as there could be more than 1 discount to apply on an item
const OFFERS: readonly Offer[] = [
  {
    itemId: "34583",
    qualifier: 3,
    name: "3 for 2 on all Beans",
    for: 2,
  },
  {
    itemId: "39845",
    qualifier: 2,
    name: "2 for Â£1 on all coke",
    price: {
      amount: 100,
      divisor: 100,
    },
  },
];

export default createSlice({
  initialState: OFFERS,
  name: "offers",
  reducers: {},
});

export const selectOffers = (state: RootState) => state.offers;
