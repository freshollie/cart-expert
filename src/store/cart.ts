import { createSelector, createSlice, PayloadAction } from "@reduxjs/toolkit";
import { RootState } from ".";
import { selectItemsLookup } from "./items";

const initialState = {} as Record<string, number>;
export default createSlice({
  name: "cart",
  initialState,
  reducers: {
    add: (
      state,
      { payload: { id, amount = 1 } }: PayloadAction<{ id: string; amount?: number }>
    ) => ({
      ...state,
      [id]: (state[id] ?? 0) + amount,
    }),
    remove: (
      state,
      { payload: { id, amount = 1 } }: PayloadAction<{ id: string; amount?: number }>
    ) => ({
      ...state,
      [id]: (state[id] ?? 0) - amount > 0 ? state[id]! - amount : 0,
    }),
    reset: () => initialState,
  },
});

export const selectCartAmount = (itemId: string) => (state: RootState) =>
  state.cart[itemId] ?? 0;

export const selectCartItems = createSelector(
  [(state: RootState) => state.cart, selectItemsLookup],
  (cart, items) =>
    Object.entries(cart)
      .map(([id, amount]) => ({ item: items[id], amount }))
      .filter(({ item, amount }) => item && amount > 0)
      .sort((item1, item2) => item1.item.id.localeCompare(item2.item.id))
);
