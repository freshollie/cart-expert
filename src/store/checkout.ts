import { createSelector } from "@reduxjs/toolkit";
import { isPriceOffer, Offer, Price } from "../types";
import { selectCartItems } from "./cart";
import { selectOffers } from "./offers";

export type Discount = { offer: Offer, savings: Price };
type CheckoutDetails = {
  total: Price;
  subTotal: Price;
  discounts: Discount[];
};

export const selectCheckoutDetails = createSelector(
  [selectCartItems, selectOffers],
  (cart, offers): CheckoutDetails => {
    // find all the discounts which apply to this cart
    const discounts = offers
      .map((offer) => {
        const cartItem = cart.find(({ item }) => offer.itemId === item.id);
        const numDiscounts = Math.floor(
          (cartItem?.amount ?? 0) / offer.qualifier
        );

        if (cartItem && numDiscounts > 0) {
          return new Array(numDiscounts).fill(0).map(() => ({
            offer,
            savings: {
              amount:
                offer.qualifier * cartItem.item.price.amount -
                (isPriceOffer(offer)
                  ? offer.price.amount
                  : offer.for * cartItem.item.price.amount),
              divisor: 100,
            },
          }));
        }
        return [];
      })
      .flat();

    // create totals and subtotals
    const subTotal = {
      amount: cart.reduce(
        (acc, { amount, item }) => acc + amount * item.price.amount,
        0
      ),
      divisor: 100,
    };
    const total = {
      amount:
        subTotal.amount -
        discounts.reduce((acc, discount) => acc + discount.savings.amount, 0),
      divisor: 100,
    };

    return {
      discounts,
      subTotal,
      total,
    };
  }
);
