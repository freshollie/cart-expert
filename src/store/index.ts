import cart from "./cart";
import items from "./items";
import { configureStore } from "@reduxjs/toolkit";
import offers from "./offers";

const store = configureStore({
  reducer: {
    cart: cart.reducer,
    items: items.reducer,
    offers: offers.reducer
  },
});

export type RootState = ReturnType<typeof store.getState>;

export default store;