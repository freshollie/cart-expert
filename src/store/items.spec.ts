import * as items from "./items";
import store from ".";

describe("items", () => {
  it("should provide the list of available items", () => {
    const list = items.selectAllItems(store.getState());
    expect(list.length).toBeGreaterThanOrEqual(3);
    expect(list).toMatchSnapshot();
  });

  it("should allow items to be quiried by id", () => {
    const item = items.selectAllItems(store.getState())[2];
    expect(items.selectItembyId(store.getState(), { id: item.id})).toBe(item);
  })
});
