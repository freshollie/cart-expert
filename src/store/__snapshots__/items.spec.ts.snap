// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`items should provide the list of available items 1`] = `
Array [
  Object {
    "id": "34583",
    "name": "Beans",
    "price": Object {
      "amount": 50,
      "divisor": 100,
    },
  },
  Object {
    "id": "39845",
    "name": "Coke",
    "price": Object {
      "amount": 70,
      "divisor": 100,
    },
  },
  Object {
    "id": "32844",
    "name": "Oranges",
    "price": Object {
      "amount": 199,
      "divisor": 100,
      "perKg": true,
    },
  },
  Object {
    "id": "34242",
    "name": "Bananas",
    "price": Object {
      "amount": 199,
      "divisor": 100,
      "perKg": true,
    },
  },
]
`;
